orderapi/src/main/java/fi/hel/verkkokauppa/order/service/invoice/InvoiceService.java

```diff
@@ -15,14 +18,27 @@ public class InvoiceService {
 
     final InvoiceMapper invoiceMapper;
 
+    final IncrementId incrementId;
+
     @Autowired
-    public InvoiceService(OrderRepository orderRepository, InvoiceMapper invoiceMapper) {
+    public InvoiceService(OrderRepository orderRepository, InvoiceMapper invoiceMapper, IncrementId incrementId) {
         this.orderRepository = orderRepository;
         this.invoiceMapper = invoiceMapper;
+        this.incrementId = incrementId;
     }
 
-    public Order saveInvoiceToOrder(InvoiceDto invoice, Order order) {
-        order.setInvoice(invoiceMapper.fromDto(invoice));
+    public Order saveInvoiceToOrder(InvoiceDto invoiceDto, Order order) {
+        Invoice invoice = invoiceMapper.fromDto(invoiceDto);
+        invoice.setInvoiceId(order.getInvoice() == null || order.getInvoice().getInvoiceId().isEmpty()
+                ? generateInvoiceId()
+                : order.getInvoice().getInvoiceId()
+        );
+        order.setInvoice(invoice);
         return orderRepository.save(order);
     }
+
+    protected String generateInvoiceId(){
+        Long invoiceIncrement = incrementId.generateInvoiceIncrementId();
+        return TypePrefix.INVOICE.number + String.format("%09d", invoiceIncrement);
```

Comment:
I think you can remove .number from this call, because enum should return same value.

Reply:
The .number call is necessary here. If there is only TypePrefix.INVOICE reference, then it will return the INVOICE enum instance.

Reply:
Just what Martin said.

Reply:
Oh then this is ok! Good explanations thanks!